### PR TITLE
#688: thread pools documentation

### PR DESCRIPTION
**Pull request by lorban**
Please note that my #165 PR is required before this can be merged.

### CODE REVIEW

File: docs/src/docs/asciidoc/user/thread-pools.adoc [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ * There can be any number of pools. Add them via `addPool`.
+
+=== Using the configured thread pools
+
+ * List of services making use of `ExecutionService`:
+   ** Disk store: disk writes are performed asynchronously.
+     `OffHeapDiskStoreConfiguration` and `OffHeapDiskStoreProviderConfiguration`
+   ** Write Behind: `CacheLoaderWriter` write tasks happen asynchronously.
+     `DefaultWriteBehindConfiguration` and `WriteBehindProviderConfiguration`
+   ** Eventing: events can be configured to be sent asynchronously.
```
Comment (ljacomet) [inline docs/src/docs/asciidoc/user/thread-pools.adoc]: Even synchronous eventing uses thread pool

File: docs/src/docs/asciidoc/user/thread-pools.adoc [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ * `PooledExecutionService` Keeps a configurable set of thread pools and divides them to service all executor service
+   requests. This implementation must be configured with a `PooledExecutionServiceConfiguration` when needed.
+
+=== Configuring `PooledExecutionService`
+
+ * Done via `PooledExecutionServiceConfiguration`
+ * There can be at most one default pool, add it via `addDefaultPool`. Its name does not matter.
+ * There can be any number of pools. Add them via `addPool`.
+
+=== Using the configured thread pools
```
Comment (ljacomet) [inline docs/src/docs/asciidoc/user/thread-pools.adoc]: This section needs to be fleshed out a bit better. Would be good to have more context on why you add class names here.

File: docs/src/docs/asciidoc/user/thread-pools.adoc [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+==== Write Behind
+
+[source,java,indent=0]
+----
+include::{sourcedir}/impl/src/test/java/org/ehcache/docs/ThreadPools.java[tag=writeBehind]
+----
+
+<1> Configure the thread pools. Note that the default one (`dflt`) is required for the events even when no event
+    listener is configured.
+<2> Use a `WriteBehindProviderConfiguration` to tell the `CacheManagerBuilder` to use a default thread pool for all
```
Comment (ljacomet) [inline docs/src/docs/asciidoc/user/thread-pools.adoc]: this hints at a missing method on `CacheManagerBuilder`

File: docs/src/docs/asciidoc/user/thread-pools.adoc [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+==== Disk store
+
+[source,java,indent=0]
+----
+include::{sourcedir}/impl/src/test/java/org/ehcache/docs/ThreadPools.java[tag=diskStore]
+----
+
+<1> Configure the thread pools. Note that the default one (`dflt`) is required for the events even when no event
+    listener is configured.
+<2> Use a `OffHeapDiskStoreProviderConfiguration` to tell the `CacheManagerBuilder` to use a default thread pool for
```
Comment (ljacomet) [inline docs/src/docs/asciidoc/user/thread-pools.adoc]: this hints at a missing method on CacheManagerBuilder

File: docs/src/docs/asciidoc/user/thread-pools.adoc [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    all disk stores that don't explicitly specify one.
+<3> Add a `OffHeapDiskStoreConfiguration` to tell the cache to use a specific thread pool for its disk store.
```
Comment (ljacomet) [inline docs/src/docs/asciidoc/user/thread-pools.adoc]: this hints at a missing method on `CacheConfigurationBuilder`

File: docs/src/docs/asciidoc/user/thread-pools.adoc (hunk 1 above)
Comment (lorban) [inline docs/src/docs/asciidoc/user/thread-pools.adoc]: You're right, I will rephrase that.

File: docs/src/docs/asciidoc/user/thread-pools.adoc (hunk 3 above)
Comment (lorban) [inline docs/src/docs/asciidoc/user/thread-pools.adoc]: Indeed, will do.

File: docs/src/docs/asciidoc/user/thread-pools.adoc (hunk 4 above)
Comment (lorban) [inline docs/src/docs/asciidoc/user/thread-pools.adoc]: Indeed, will do.

File: docs/src/docs/asciidoc/user/thread-pools.adoc (hunk 5 above)
Comment (lorban) [inline docs/src/docs/asciidoc/user/thread-pools.adoc]: Indeed, will do.

File: docs/src/docs/asciidoc/user/thread-pools.adoc [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+<3> Tell the cache to use a specific thread pool for its disk store.
+
+==== Write Behind
+
+[source,java,indent=0]
+----
+include::{sourcedir}/impl/src/test/java/org/ehcache/docs/ThreadPools.java[tag=writeBehind]
+----
+
+<1> Configure the thread pools. Note that the default one (`dflt`) is required for the events even when no event
```
Comment (chrisdennis) [inline docs/src/docs/asciidoc/user/thread-pools.adoc]: Do we really need to repeat this callout?

**Comment by ljacomet** [comment]
Reading the doc and reading the comment from @chrisdennis about repeated warning in call out, I am wondering if we should not simply bite the bullet and make having a default thread pool mandatory. I have the impression that while not perfect it helps the user make a proper config faster.

**Comment by lorban** [comment]
I'm not fond of the idea of making the default thread pool mandatory.
When a user want total control of the thread pools, not configuring a default one is an easy way to make sure that you haven't overlooked anything as you'll get an exception during initialization.

On the other hand, I do agree that it's quite an annoyance to have to configure a default thread pool for events even when you don't even register a listener.

**Pull request merged**